### DISCUSSION TITLE
Where should i set the culture for localization?

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#12770] [Q&A] [answered]
The [Documentation](https://docs.avaloniaui.net/docs/next/guides/implementation-guides/localizing#set-the-culture) seems to state that developers have to use the following to set the application culture :
```cs
Assets.Resources.Culture = culture;
```
But that doesn't seem to be right as `Assets` isn't defined anywhere, even in source generated files.

So what am i supposed to do? set the thread's culture?

The same issue is occurring when trying to get those localized strings in the views

**Comment by maxkatz6** [comment]
It's a question about common .NET resource generators, not avalonia. I.e. you can find more information globally.

But answering your question, generator creates namespace/classes depending on the folder structure. I.e. if you have your "resx" files inside of the Assets folder, then it will generate Assets namespace. In other words, instead of "Assets" it might be other folder.

I will adjust documentation to include this comment.

**Reply by Mrcubix** [reply]
Couldn't get it to work, had my localizations in Assets/Localizations.
Used the same source gen as the project that was provided (with changes to paths of course), nothing is generated, 
nothing is said about the generation.

The namespace in question being `[Root].Assets.Localizations` Doesn't seems to exist after building and therefore doesn't contains any class.

**Reply by timunie** [reply]
iirc it was `CultureInfo.CurrentUiCulture` which one needs to set, but I may be wrong though.

**Reply by Mrcubix** [reply]
Managed to get the static properties to show up in vscode by manually generating the source files from the resx, though i am still unable to use those in axaml.

Tried
```axaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:locals="clr-namespace:ColorTiles.Assets.Localizations"
...  />

<TextBlock Text="{x:Static locals:Resources.ScoreLabel}"  />
```
Got
```err
Avalonia error AVLN:0004: Unable to resolve ScoreLabel as static field, property, constant or enum value
```

**Reply by IanRawley** [reply]
I have a Localization directory in my project root with AppStrings.resx & AppStrings.ja-JP.resx (for Japanese localization).  To use them I have:
```
xmlns:L="clr-namespace:ProjectRoot.Localization"
```
and use them with 
```
HeaderText="{x:Static L:AppStrings.Label_Name}"
```

Make sure the resx files are set to Public (in VS2022 a dropdown in the top right of the resx editing view, I don't know about VScode) and that should work.  Probably also make sure that the resx Build Action is Embedded Resource.

**Reply by Mrcubix** [reply]
The code in my last reply is what i'm basically using, only thing is that i don't use Visual Bloat, but vscode, which could be the reason for the resx sourcegen not working.

As it is right now, Visual Bloat of not, it should be working, yet Avalonia isn't finding those static resources.

**Reply by timunie** [reply]
in the resx tool, make sure that generated code is public. It's internal by default which will not work with Avalonia well. 

----

Another idea: How huge is your App? Probably you can consider to use `DynamicResources` instead if the app is small or medium sized, as this will support dynamic language switching. I can share a sample if you are interested in.

**Reply by Mrcubix** [reply]
> in the resx tool, make sure that generated code is public. It's internal by default which will not work with Avalonia well.

Seems like it does generate with internal, fixed that part, now getting this exception :

```cs
An unhandled exception of type 'System.Resources.MissingManifestResourceException' occurred in System.Private.CoreLib.dll: 'Could not find the resource "Localizations.Resources.resources" among the resources "ColorTiles.Assets.Localizations.Resources.resources", "!AvaloniaResources" embedded in the assembly "ColorTiles", nor among the resources in any satellite assemblies for the specified culture. Perhaps the resources were embedded with an incorrect name.'
```

Which i managed to fix by fixing the namespace the generator used wrong twice

![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/1cc9eb38-c28d-4762-a45d-da661a76aee0)
![image](https://github.com/AvaloniaUI/Avalonia/assets/39861216/ddeb7eb4-ace0-4e4d-8fcf-55da40744cc4)

Which results in a properly working Fr*nch translation

![Image](https://i.imgur.com/KVcxiqT.png)

> Another idea: How huge is your App? Probably you can consider to use `DynamicResources` instead if the app is small or medium sized, as this will support dynamic language switching. I can share a sample if you are interested in.

Game is really simple so i could try your suggestion if it can make this process simpler.

**Reply by Mrcubix** [reply, marked as answer]
After writing that message, i found an actually working way of making localizations work anywhere, including vscode.

In the provided [Sample](https://github.com/AvaloniaUI/AvaloniaUI.QuickGuides/blob/main/Localization/Localization.csproj), The generator is never run by Dotnet CLI, so my guess is that it only works in Visual Bloat 20xx.

```csproj
<ItemGroup>
    <EmbeddedResource Update="Assets\Localizations\Resources.resx">
      <Generator>MSBuild:Compile</Generator>
      <LastGenOutput>%(FileName).Designer.cs</LastGenOutput>
      <publicClass>true</publicClass>
      <StronglyTypedFileName>$(IntermediateOutputPath)\%(FileName).Designer.cs</StronglyTypedFileName>
      <StronglyTypedLanguage>CSharp</StronglyTypedLanguage>
      <StronglyTypedNamespace>ColorTiles.Assets.Localizations</StronglyTypedNamespace>
      <StronglyTypedClassName>%(FileName)</StronglyTypedClassName>
    </EmbeddedResource>
</ItemGroup>
```

The `publicClass` (not to be confused with the `PublicClass` property from the `GenerateResource` class) argument is important as it is passed to resgen and makes all necessary accessors public instead of internal.
The generator code in the [Sample](https://github.com/AvaloniaUI/AvaloniaUI.QuickGuides/blob/main/Localization/Localization.csproj) should probably be replaced.
